invenio_rdm_records/services/customizations.py

```diff
@@ -41,57 +41,133 @@ def customize(cls, opts):
         return _make_cls(cls, attrs) if attrs else cls
 
 
-class FileConfigMixin:
-    """Shared customization for file service configs."""
+class FromConfig:
+    """Data descriptor to connect config with application configuration.
 
-    @classmethod
-    def customize(cls, permission_policy=None):
-        """Class method to customize the config for an instance."""
-        attrs = {}
-        # Permission policy
-        if permission_policy:
-            attrs['permission_policy_cls'] = permission_policy
-        # Create the config class
-        return _make_cls(cls, attrs) if attrs else cls
+    See https://docs.python.org/3/howto/descriptor.html .
 
+    .. code-block:: python
 
-class RecordConfigMixin:
-    """Shared customization for record service configs."""
+        # service/config.py
+        class ServiceConfig:
+            foo = FromConfig("FOO", default=1)
 
-    @classmethod
-    def customize(cls, permission_policy=None, pid_providers=None, pids=None,
-                  doi_enabled=False, **kwargs):
-        """Class method to customize the config for an instance."""
-        attrs = {}
+        # config.py
+        FOO = 2
+
+        # ext.py
+        c = ServiceConfig.build()
+        c.foo  # 2
+    """
+
+    def __init__(self, config_key, default=None, import_string=False):
+        """Constructor for data descriptor."""
+        self.config_key = config_key
+        self.default = default
+        self.import_string = import_string
+
+    def __get__(self, obj, objtype=None):
+        """Return value that was grafted on obj (descriptor protocol)."""
+        if self.import_string:
+            return load_or_import_from_config(
+                key=self.config_key, default=self.default
+            )
+        else:
+            return current_app.config.get(self.config_key, self.default)
+
+    def __set_name__(self, owner, name):
+        """Store name of grafted field (descriptor protocol)."""
+        # If we want to allow setting it we can implement this.
+        pass
+
+    def __set__(self, obj, value):
+        """Set value on grafted_field of obj (descriptor protocol)."""
+        # If we want to allow setting it we can implement this.
+        pass
+
+
+class FromConfigPIDsProviders:
+    """Data descriptor for pid providers configuration."""
+
+    def __get__(self, obj, objtype=None):
```

Comment:
one of the effects of using descriptors here would be that the service config properties would be loaded lazily (and in fact, the `__get__(...)` would be re-executed on each access) instead of eagerly as it was until now

this means that we need to be careful regarding how heavy the operation of fetching and building the config properties is (not an issue here, i suppose) and maybe do some caching
but this kind of behavior would also allow for `app.config` to change during run-time, and have the changes take effect without restarting the application :thinking: 

just a rough thought:
if it would become a thing to populate parts of `app.config` via Vault (a secrets management engine) during runtime, this kind of approach could come in handy with [Vault's dynamic secrets](https://learn.hashicorp.com/tutorials/vault/getting-started-dynamic-secrets)

Reply:
> one of the effects of using descriptors here would be that the service config properties would be loaded lazily (and in fact, the __get__(...) would be re-executed on each access) instead of eagerly as it was until now

Yes, I was thinking about that too. All configs are memory based operations (no disk or network access), so I am not worried about performance at all. The dynamic nature of it is a nice added benefit for the future. There would be some work to be done in [invenio-config](https://github.com/inveniosoftware/invenio-config) to really get there though.

Vaults sounds cool! Hashicorp stuff is really on point.